apisix/balancer/ewma.lua

```diff
@@ -0,0 +1,240 @@
+    end
+
+    success, err, forcible = shm_ewma:set(upstream, ewma)
+    if not success then
+        core.log.error("balancer_ewma:set failed ", err)
+    end
+    if forcible then
+        core.log.warn("balancer_ewma:set valid items forcibly overwritten")
+    end
+end
+
+
+local function get_or_update_ewma(upstream, rtt, update)
+    local lock_ok, err = nil
+    if update then
+        lock_ok, err = lock(upstream)
+    end
+    local ewma = shm_ewma:get(upstream) or 0
+    if not lock_ok then
+        return ewma, err
+    end
+
+    local now = ngx_now()
+    local last_touched_at = shm_last_touched_at:get(upstream) or 0
+    ewma = decay_ewma(ewma, last_touched_at, rtt, now)
+
+    if not update then
+        return ewma
+    end
+
+    store_stats(upstream, ewma, now)
+
+    unlock()
+
+    return ewma
+end
+
+
+local function score(upstream)
+    -- Original implementation used names
+    -- Endpoints don't have names, so passing in IP:Port as key instead
+    local upstream_name = upstream.address .. ":" .. upstream.port
+    return get_or_update_ewma(upstream_name, 0, false)
+end
+
+
+-- implementation similar to https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
+-- or https://en.wikipedia.org/wiki/Random_permutation
+-- loop from 1 .. k
+-- pick a random value r from the remaining set of unpicked values (i .. n)
+-- swap the value at position i with the value at position r
+local function shuffle_peers(peers, k)
+    for i = 1, k do
+        local rand_index = math.random(i, #peers)
+        peers[i], peers[rand_index] = peers[rand_index], peers[i]
+    end
+end
+
+
+local function pick_and_score(peers, k)
+    shuffle_peers(peers, k)
+    local lowest_score_index = 1
+    local lowest_score = score(peers[lowest_score_index])
+    for i = 2, k do
+        local new_score = score(peers[i])
+        if new_score < lowest_score then
+            lowest_score_index, lowest_score = i, new_score
+        end
+    end
+
+    return peers[lowest_score_index], lowest_score
+end
+
+
+local function _trans_format(t1)
+    -- trans
+    --{"1.2.3.4:80":100,"5.6.7.8:8080":100}
+    -- into
+    -- [{"address":"1.2.3.4","port":"80"},{"address":"5.6.7.8","port":"8080"}]
+    local t2 = {}
+    local addr, port, err
+
+    for k,_ in pairs(t1) do
+        addr, port, err = core.utils.parse_addr(k)
+        if not err then
+            core.table.insert(t2, {address = addr, port = tostring(port)})
+        else
+            core.log.error('parse_addr error: ', k, err)
+        end
+    end
+
+    return next(t2) and t2 or nil
+end
+
+
+local function _ewma_find(up_nodes)
+    local peers
+    local endpoint
+    local err
+
+    if not ewma_lock then
+        ewma_lock, ewma_lock_err = resty_lock:new("balancer_ewma_locks",
+                                                  {timeout = 0, exptime = 0.1})
+    end
+    if not ewma_lock then
+        return nil, ewma_lock_err
+    end
+    peers = _trans_format(up_nodes)
```

Comment:
I think we can use a `lrucache` to get the `peers`, avoid to generate `peer` for every time.